pkg/ipam/networks.go

```diff
@@ -16,95 +16,102 @@ package ipam
-func (lipam *LiqoIPAM) acquireNetwork(cidr string, preAllocated uint, immutable bool) (string, error) {
+// networkAcquireSpecific acquires a network with a specific prefix.
+// If the network is already allocated, it returns an error.
+func (lipam *LiqoIPAM) networkAcquireSpecific(cidr string) (string, error) {
 	lipam.mutex.Lock()
 	defer lipam.mutex.Unlock()
 
-	// TODO: implement real network acquire logic
-	_ = immutable
-	if lipam.cacheNetworks == nil {
-		lipam.cacheNetworks = make(map[string]networkInfo)
-	}
-	nw := network{
-		cidr:         cidr,
-		preAllocated: preAllocated,
+	prefix, err := netip.ParsePrefix(cidr)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse CIDR %q: %w", cidr, err)
 	}
-	lipam.cacheNetworks[nw.String()] = networkInfo{
-		network:           nw,
-		creationTimestamp: time.Now(),
+
+	result := lipam.IpamCore.NetworkAcquireWithPrefix(prefix)
+	if result == nil {
+		return "", fmt.Errorf("failed to reserve specific network %q", cidr)
 	}
 
-	klog.Infof("Acquired network %q", nw.cidr)
-	return nw.cidr, nil
+	klog.Infof("Acquired specific network %q -> %q", cidr, result.String())
+
+	if lipam.opts.GraphvizEnabled {
+		return result.String(), lipam.IpamCore.ToGraphviz()
+	}
+	return result.String(), nil
 }
 
-// freeNetwork frees a network, removing it from the cache.
-func (lipam *LiqoIPAM) freeNetwork(nw network) {
+// networkRelease frees a network, removing it from the cache.
+func (lipam *LiqoIPAM) networkRelease(cidr string) error {
 	lipam.mutex.Lock()
 	defer lipam.mutex.Unlock()
 
-	// TODO: implement real network free logic
-	delete(lipam.cacheNetworks, nw.String())
-	klog.Infof("Freed network %q", nw.cidr)
+	prefix, err := netip.ParsePrefix(cidr)
+	if err != nil {
+		return fmt.Errorf("failed to parse CIDR %q: %w", cidr, err)
+	}
+
+	result := lipam.IpamCore.NetworkRelease(prefix)
+	if result == nil {
+		return fmt.Errorf("failed to free network %q", cidr)
+	}
+	klog.Infof("Freed network %q", cidr)
+
+	if lipam.opts.GraphvizEnabled {
+		return lipam.IpamCore.ToGraphviz()
+	}
+	return nil
 }
 
-// isNetworkAvailable checks if a network is available.
-func (lipam *LiqoIPAM) isNetworkAvailable(nw network) bool {
+// networkIsAvailable checks if a network is available.
+func (lipam *LiqoIPAM) networkIsAvailable(cidr string) (bool, error) {
 	lipam.mutex.Lock()
 	defer lipam.mutex.Unlock()
 
-	// TODO: implement real network availability check logic
-	if lipam.cacheNetworks == nil {
-		return true
+	prefix, err := netip.ParsePrefix(cidr)
+	if err != nil {
+		return false, fmt.Errorf("failed to parse CIDR %q: %w", cidr, err)
 	}
-	_, ok := lipam.cacheNetworks[nw.String()]
 
-	return ok
+	return !lipam.IpamCore.NetworkIsAvailable(prefix), nil
 }
 
-func listNetworksOnCluster(ctx context.Context, cl client.Client) ([]network, error) {
-	var nets []network
+func (lipam *LiqoIPAM) listNetworksOnCluster(ctx context.Context, cl client.Client) ([]netip.Prefix, error) {
```

Comment:
Here the original function returned the network struct which also containes the preAllocated field. We can not remove it since we need this info the initialize and sync routines